File: practica.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define SIZE 10000

void swap(int *xp, int *yp) {
    int temp = *xp;
    *xp = *yp;
    *yp = temp;
}

void simple(int arr[], int n) {
    int i, j, min_idx;
    for (i = 0; i < n-1; i++) {
        min_idx = i;
        for (j = i+1; j < n; j++)
            if (arr[j] < arr[min_idx])
                min_idx = j;
        swap(&arr[min_idx], &arr[i]);
    }
}

void bubble(int arr[], int n) {
    int i, j;
    for (i = 0; i < n-1; i++)        
        for (j = 0; j < n-i-1; j++)
            if (arr[j] > arr[j+1])
                swap(&arr[j], &arr[j+1]);
}

void insertion(int arr[], int n) {
    int i, key, j;
    for (i = 1; i < n; i++) {
        key = arr[i];
        j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

double timer(int arr[], int n, void (*func)(int*, int)) {
    clock_t start, end;
    double cpu_time_used;
    start = clock();
    func(arr, n);
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    return cpu_time_used;
}

void fillArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        arr[i] = rand() % 10000;
    }
}

int main() {
    int arr[SIZE];
    double simple_time, bubble_time, insertion_time, best_time;
    srand(time(0)); // инициализация

    fillArray(arr, SIZE);
    simple_time = timer(arr, SIZE, simple);
    printf("Simple sort took %f seconds.\n", simple_time);

    fillArray(arr, SIZE);
    bubble_time = timer(arr, SIZE, bubble);
    printf("Bubble sort took %f seconds.\n", bubble_time);

    fillArray(arr, SIZE);
    insertion_time = timer(arr, SIZE, insertion);
    printf("Insertion sort took %f seconds.\n", insertion_time);

    best_time = simple_time < bubble_time ? (simple_time < insertion_time ? simple_time : insertion_time) : (bubble_time < insertion_time ? bubble_time : insertion_time);
    printf("Best time: %f seconds.\n", best_time);

    return 0;
}

File: practica.py
import time
import random

def simple(arr):
    # Simple selection sort
    n = len(arr)
    for i in range(n):
        min_idx = i
        for j in range(i+1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr

def bubble(arr):
    # Bubble sort
    n = len(arr)
    for i in range(n):
        for j in range(0, n-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr

def insertion(arr):
    # Insertion sort
    n = len(arr)
    for i in range(1, n):
        key = arr[i]
        j = i-1
        while j >=0 and key < arr[j]:
            arr[j+1] = arr[j]
            j -= 1
        arr[j+1] = key
    return arr


def timer(arr, func):
    # Counts time of a sorting function
    start = time.time()
    result = func(arr.copy())  
    end = time.time()
    print(f"{func.__name__} потратила {end - start} секунд.")
    return result, end-start

def main():
    arr = [random.randint(0, 10000) for _ in range(10)]
    print(f"Исходный массив: {arr}")
    simple_sort_result, simple_sort = timer(arr, simple)
    bubble_sort_result, bubble_sort = timer(arr, bubble)
    insertion_sort_result, insertion_sort = timer(arr, insertion)
    print(simple_sort_result, bubble_sort_result, insertion_sort_result)
    best = min(simple_sort, bubble_sort, insertion_sort)
    print(f"Наименьший результат: {best} у функции {'simple' if best == simple_sort else 'bubble' if best == bubble_sort else 'insertion'}")

if __name__ == "__main__":
    main()